include/seqan3/range/view/single_pass_input.hpp

```diff
@@ -0,0 +1,402 @@
+    template <typename view_t>
+    friend class single_pass_input_iterator;
+
+    //!\brief The underlying sentinel state.
+    sentinel_t sentinel{};
+
+public:
+    /*!\name Constructor, destructor and assignment
+     * \{
+     */
+    //!\brief Default constructor.
+    single_pass_input_sentinel() = default;
+    //!\brief Copy constructor.
+    constexpr single_pass_input_sentinel(single_pass_input_sentinel const &) = default;
+    //!\brief Move constructor.
+    constexpr single_pass_input_sentinel(single_pass_input_sentinel &&) = default;
+    //!\brief Copy assignment.
+    constexpr single_pass_input_sentinel & operator=(single_pass_input_sentinel const &) = default;
+    //!\brief Move assignment.
+    constexpr single_pass_input_sentinel & operator=(single_pass_input_sentinel &&) = default;
+    //!\brief Destructor.
+    ~single_pass_input_sentinel() = default;
+
+    //!\brief Construction from underlying sentinel.
+    single_pass_input_sentinel(sentinel_t _s) : sentinel{_s}
+    {}
+    //!\}
+
+    /*!\name Comparison
+     * \{
+     */
+    //!\brief Comparison with iterator type.
+    template <typename view_t>
+    bool operator==(single_pass_input_iterator<view_t> const & rhs) const noexcept
+    {
+        return sentinel == rhs.cached();
+    }
+
+    //!\brief Comparison with iterator type.
+    template <typename view_t>
+    bool operator!=(single_pass_input_iterator<view_t> const & rhs) const noexcept
+    {
+        return !(*this == rhs);
+    }
+    //!/}
+};
+} // seqan3::detail
+
+//-----------------------------------------------------------------------------
+// Implementation of single pass input view.
+//-----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+
+/*!\brief Adds single_pass_input behavior to the underlying range.
+ * \tparam urng_t The underlying range type.
+ * \implements input_range_concept
+ * \ingroup view
+ */
+template <input_range_concept urng_t>
+class single_pass_input_view : public detail::view_base
+{
+private:
+
+    //!\brief Friend declaration for seqan3::detail::single_pass_input_iterator.
+    template <typename view_t>
+    friend class single_pass_input_iterator;
+
+    //!\brief The pure range type without any reference type.
+    using pure_range_t       = std::remove_reference_t<urng_t>;
+    //!\brief The iterator type for the underlying range.
+    using urng_iterator_type = iterator_t<pure_range_t>;
+    //!\brief The sentinel type for the underlying range.
+    using sentinel           = single_pass_input_sentinel<sentinel_t<pure_range_t>>;
+
+    //!\brief The internal state.
+    struct state
+    {
+        //!\brief The underlying range.
+        urng_t             urng;
+        //!\brief The cached iterator of the underlying range.
+        urng_iterator_type cached_urng_iter{};
+    };
+
+    //!\brief Shared pointer of the data model.
+    std::shared_ptr<state> state_ptr{};
```

Comment:
we don't include `_ptr` anywhere else, but we do include `_t` or `_type` in all our member type definitions, except where the STL doesn't. So I would propose:
```cpp
struct state_t
{
// ...
};
std::shared_ptr<state_t> state{};
```
Or `data_members_t` and `data_members` like in the tutorial!

Reply:
I can't follow your intentions. This is an internal name for variable not exposed to the user. And it is expressive in what it means. Just because other's don't use that style (yet), it doesn't mean that we can't do this. I totally agree on public interfaces and keeping them STL-like, but IMHO this get's a little out of scope regarding our design goals.

Reply:
For the general argument see below.

For the concrete situation: when I read `state_ptr` I assumed it was a type. All other things that end in `_ptr` are types, like `shared_ptr` in the same line. On the other hand `state` doesn't look like a type at all.
Yes, it's private and it might all be considered :nail_care:, but in contrast to SeqAn2 and to other styles, it is much more difficult now to discern types and values. I think it would be important to state early that our template parameters, our local typedefs and our local type definitions always end in `_t` or `_type`, it just helps reading the code, or not?

Reply:
Well, the suffix `_t[ype]` is used to distinguish between a generic type or a typedef from a specific class name. 
So `state` is perfectly valid without `_t`, because I also don't call it `class suffix_tree_t{};` or something. The `_ptr` is a visual suffix to tell the developer/user that the underlying type in fact is a pointer to an object. So I can't use member functions directly on it but have to dereference before. No matter if it is a raw- fancy- or smart-pointer.